src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,153 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            Node<T> node = new Node<>();
+            node.value = value;
+            head = node;
+            tail = node;
+
+        } else {
+            Node<T> node = new Node<>();
+            node.value = value;
+            node.previous = tail;
+            tail.next = node;
+            tail = node;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (index == size) {
+            Node<T> previousTail = tail;
+            tail = new Node<>();
+            tail.previous = previousTail;
+            tail.value = value;
+            if (previousTail == null) {
+                head = tail;
+            } else {
+                previousTail.next = tail;
+            }
+        } else {
+            Node<T> nodeBeforeInput = getNode(index);
+            Node<T> oldPrevious = nodeBeforeInput.previous;
+            Node<T> inputNode = new Node<>();
+            inputNode.previous = oldPrevious;
+            inputNode.value = value;
+            inputNode.next = nodeBeforeInput;
+            nodeBeforeInput.previous = inputNode;
+            if (oldPrevious == null) {
+                head = inputNode;
+            } else {
+                oldPrevious.next = inputNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T previousValue = node.value;
+        node.value = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> target = getNode(index);
+        removingAction(target);
+        return target.value;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> node = head;
+        while (node != null) {
+            if (node.value == t || node.value.equals(t)) {
```

Comment:
there is a risk of NPE

Reply:
done additional null-checking at the beginning of the method
is it enough or need to do something else here to avoid npe?

Reply:
replaced while loop to for with size-limit
no risk of npe now, i guess